File: src/util/fo_fs.f90
module fo_fs
    !! Shell-free filesystem operations. Every routine here calls a libc
    !! primitive (via fo_fs.c) or pure Fortran I/O, never execute_command_line,
    !! so nothing forks /bin/sh and nothing is corrupted from an OpenMP region.
    use, intrinsic :: iso_c_binding, only: c_char, c_int, c_null_char
    implicit none
    private

    public :: fs_remove_tree, fs_remove_file, fs_make_dir
    public :: fs_delete_suffix, fs_append_file, fs_write_text
    public :: fs_collect_files, fs_collect_mod_dirs
    public :: fs_mkdir_excl, fs_sleep_ms, fs_pid_alive
    public :: fs_copy_exec, fs_rename

    interface
        integer(c_int) function fo_c_rm_rf(path) bind(C, name='fo_c_rm_rf')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: path(*)
        end function fo_c_rm_rf

        integer(c_int) function fo_c_rm_file(path) bind(C, name='fo_c_rm_file')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: path(*)
        end function fo_c_rm_file

        integer(c_int) function fo_c_mkdir_p(path) bind(C, name='fo_c_mkdir_p')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: path(*)
        end function fo_c_mkdir_p

        integer(c_int) function fo_c_delete_suffix(root, suffix, recursive) &
                bind(C, name='fo_c_delete_suffix')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: root(*), suffix(*)
            integer(c_int), value :: recursive
        end function fo_c_delete_suffix

        integer(c_int) function fo_c_append_file(src, dst) &
                bind(C, name='fo_c_append_file')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: src(*), dst(*)
        end function fo_c_append_file

        integer(c_int) function fo_c_copy_exec(src, dst) &
                bind(C, name='fo_c_copy_exec')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: src(*), dst(*)
        end function fo_c_copy_exec

        integer(c_int) function fo_c_rename_path(src, dst) &
                bind(C, name='fo_c_rename_path')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: src(*), dst(*)
        end function fo_c_rename_path

        integer(c_int) function fo_c_collect_files(root, infix, suffix, &
                path_needle, recursive, out, cap) &
                bind(C, name='fo_c_collect_files')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: root(*), infix(*), suffix(*)
            character(kind=c_char), intent(in) :: path_needle(*)
            integer(c_int), value :: recursive
            character(kind=c_char), intent(out) :: out(*)
            integer(c_int), value :: cap
        end function fo_c_collect_files

        integer(c_int) function fo_c_mkdir_excl(path) &
                bind(C, name='fo_c_mkdir_excl')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: path(*)
        end function fo_c_mkdir_excl

        subroutine fo_c_sleep_ms(ms) bind(C, name='fo_c_sleep_ms')
            import :: c_int
            integer(c_int), value :: ms
        end subroutine fo_c_sleep_ms

        integer(c_int) function fo_c_pid_alive(pid) &
                bind(C, name='fo_c_pid_alive')
            import :: c_int
            integer(c_int), value :: pid
        end function fo_c_pid_alive

        integer(c_int) function fo_c_collect_mod_dirs(root, out, cap) &
                bind(C, name='fo_c_collect_mod_dirs')
            import :: c_char, c_int
            character(kind=c_char), intent(in) :: root(*)
            character(kind=c_char), intent(out) :: out(*)
            integer(c_int), value :: cap
        end function fo_c_collect_mod_dirs
    end interface

    integer, parameter :: FS_COLLECT_CAP = 262144

contains

    subroutine fs_remove_tree(path)
        !! Recursive delete (rm -rf). Missing path is success.
        character(len=*), intent(in) :: path
        integer(c_int) :: rc
        if (len_trim(path) == 0) return
        rc = fo_c_rm_rf(trim(path)//c_null_char)
    end subroutine fs_remove_tree

    subroutine fs_remove_file(path)
        !! Delete one file (rm -f). Missing file is success.
        character(len=*), intent(in) :: path
        integer(c_int) :: rc
        if (len_trim(path) == 0) return
        rc = fo_c_rm_file(trim(path)//c_null_char)
    end subroutine fs_remove_file

    subroutine fs_make_dir(path)
        !! Create path and missing parents (mkdir -p).
        character(len=*), intent(in) :: path
        integer(c_int) :: rc
        if (len_trim(path) == 0) return
        rc = fo_c_mkdir_p(trim(path)//c_null_char)
    end subroutine fs_make_dir

    subroutine fs_delete_suffix(root, suffix, recursive)
        !! Delete files under root whose name ends with suffix (find -delete).
        character(len=*), intent(in) :: root, suffix
        logical, intent(in) :: recursive
        integer(c_int) :: rec, rc
        rec = 0
        if (recursive) rec = 1
        rc = fo_c_delete_suffix(trim(root)//c_null_char, &
            trim(suffix)//c_null_char, rec)
    end subroutine fs_delete_suffix

    subroutine fs_append_file(src, dst)
        !! Append src onto dst (cat src >> dst).
        character(len=*), intent(in) :: src, dst
        integer(c_int) :: rc
        rc = fo_c_append_file(trim(src)//c_null_char, trim(dst)//c_null_char)
    end subroutine fs_append_file

    subroutine fs_write_text(path, text)
        !! Write a single line of text to path, replacing it (printf text > path).
        character(len=*), intent(in) :: path, text
        integer :: u, ios
        open (newunit=u, file=trim(path), status='replace', action='write', &
            iostat=ios)
        if (ios /= 0) return
        write (u, '(a)') text
        close (u)
    end subroutine fs_write_text

    subroutine fs_collect_files(root, infix, suffix, path_needle, items, &
            n_items, recursive)
        !! Collect regular files under root whose basename contains infix and
        !! ends with suffix and whose path contains path_needle, into items
        !! (each a full path). Recurses unless recursive is .false. Replaces a
        !! find pipeline.
        character(len=*), intent(in) :: root, infix, suffix, path_needle
        character(len=*), intent(out) :: items(:)
        integer, intent(out) :: n_items
        logical, intent(in), optional :: recursive
        character(kind=c_char), allocatable :: buf(:)
        integer(c_int) :: rc, rec

        rec = 1
        if (present(recursive)) then
            if (.not. recursive) rec = 0
        end if
        allocate (buf(FS_COLLECT_CAP))
        rc = fo_c_collect_files(trim(root)//c_null_char, trim(infix)//c_null_char, &
            trim(suffix)//c_null_char, &
            trim(path_needle)//c_null_char, rec, buf, &
            int(FS_COLLECT_CAP, c_int))
        call unpack_buffer(buf, int(rc), items, n_items)
        call sort_items(items, n_items)
        deallocate (buf)
    end subroutine fs_collect_files

    function fs_mkdir_excl(path) result(state)
        !! Atomic exclusive mkdir used as a lock: 0 created, 1 already existed,
        !! -1 error.
        character(len=*), intent(in) :: path
        integer :: state
        state = int(fo_c_mkdir_excl(trim(path)//c_null_char))
    end function fs_mkdir_excl

    subroutine fs_sleep_ms(ms)
        !! Sleep for ms milliseconds without a shell.
        integer, intent(in) :: ms
        call fo_c_sleep_ms(int(ms, c_int))
    end subroutine fs_sleep_ms

    logical function fs_pid_alive(pid)
        !! True if a process with this pid currently exists (kill -0).
        integer, intent(in) :: pid
        fs_pid_alive = fo_c_pid_alive(int(pid, c_int)) /= 0
    end function fs_pid_alive

    integer function fs_copy_exec(src, dst) result(rc)
        !! Copy src over dst and make dst executable (0755). 0 on success.
        character(len=*), intent(in) :: src, dst
        rc = int(fo_c_copy_exec(trim(src)//c_null_char, trim(dst)//c_null_char))
    end function fs_copy_exec

    integer function fs_rename(src, dst) result(rc)
        !! Atomic rename src to dst. 0 on success.
        character(len=*), intent(in) :: src, dst
        rc = int(fo_c_rename_path(trim(src)//c_null_char, trim(dst)//c_null_char))
    end function fs_rename

    subroutine fs_collect_mod_dirs(root, items, n_items)
        !! Recursively collect unique parent directories of *.mod files under
        !! root. Replaces `find -name '*.mod' -printf '%h\n' | sort -u`.
        character(len=*), intent(in) :: root
        character(len=*), intent(out) :: items(:)
        integer, intent(out) :: n_items
        character(kind=c_char), allocatable :: buf(:)
        integer(c_int) :: rc

        allocate (buf(FS_COLLECT_CAP))
        rc = fo_c_collect_mod_dirs(trim(root)//c_null_char, buf, &
            int(FS_COLLECT_CAP, c_int))
        call unpack_buffer(buf, int(rc), items, n_items)
        call sort_items(items, n_items)
        deallocate (buf)
    end subroutine fs_collect_mod_dirs

    subroutine sort_items(items, n)
        !! Stable lexical sort so collected file lists are deterministic, the
        !! way `find | sort` was, keeping cache keys and link order stable.
        character(len=*), intent(inout) :: items(:)
        integer, intent(in) :: n
        integer :: i, j
        character(len=len(items)) :: tmp

        do i = 2, n
            tmp = items(i)
            j = i - 1
            do while (j >= 1)
                if (llt(trim(tmp), trim(items(j)))) then
                    items(j + 1) = items(j)
                    j = j - 1
                else
                    exit
                end if
            end do
            items(j + 1) = tmp
        end do
    end subroutine sort_items

    subroutine unpack_buffer(buf, rc, items, n_items)
        !! Split a NUL-separated C buffer of rc entries into items(:).
        character(kind=c_char), intent(in) :: buf(:)
        integer, intent(in) :: rc
        character(len=*), intent(out) :: items(:)
        integer, intent(out) :: n_items
        integer :: i, start, k

        n_items = 0
        if (rc <= 0) return
        start = 1
        k = 0
        do i = 1, size(buf)
            if (k >= rc .or. n_items >= size(items)) exit
            if (buf(i) == c_null_char) then
                if (i > start) then
                    n_items = n_items + 1
                    call chars_to_str(buf, start, i - 1, items(n_items))
                else
                    n_items = n_items + 1
                    items(n_items) = ''
                end if
                k = k + 1
                start = i + 1
            end if
        end do
    end subroutine unpack_buffer

    subroutine chars_to_str(buf, lo, hi, str)
        character(kind=c_char), intent(in) :: buf(:)
        integer, intent(in) :: lo, hi
        character(len=*), intent(out) :: str
        integer :: i, n

        str = ''
        n = min(hi - lo + 1, len(str))
        do i = 1, n
            str(i:i) = char(ichar(buf(lo + i - 1)))
        end do
    end subroutine chars_to_str

end module fo_fs

File: src/util/fo_fs.c
/* Filesystem primitives with no shell: replacements for the rm/mkdir/find
   shell-outs fo used to make. Every operation here is a direct libc syscall,
   so nothing forks /bin/sh and nothing is corrupted when called from an
   OpenMP parallel region. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <limits.h>

#ifndef PATH_MAX
#define PATH_MAX 4096
#endif

static int fo_has(const char *s) { return s != NULL && s[0] != '\0'; }

/* Recursively delete a file or directory tree. Missing path is success
   (mirrors rm -rf). Returns 0 on success, -1 on error. */
int fo_c_rm_rf(const char *path) {
    struct stat st;
    DIR *dir;
    struct dirent *ent;
    char child[PATH_MAX];

    if (!fo_has(path)) return 0;
    if (lstat(path, &st) != 0) {
        return (errno == ENOENT) ? 0 : -1;
    }
    if (!S_ISDIR(st.st_mode)) {
        if (unlink(path) != 0 && errno != ENOENT) return -1;
        return 0;
    }

    dir = opendir(path);
    if (dir == NULL) return -1;
    while ((ent = readdir(dir)) != NULL) {
        if (strcmp(ent->d_name, ".") == 0 || strcmp(ent->d_name, "..") == 0)
            continue;
        if (snprintf(child, sizeof(child), "%s/%s", path, ent->d_name) >=
            (int)sizeof(child)) {
            closedir(dir);
            return -1;
        }
        if (fo_c_rm_rf(child) != 0) {
            closedir(dir);
            return -1;
        }
    }
    closedir(dir);
    if (rmdir(path) != 0 && errno != ENOENT) return -1;
    return 0;
}

/* Delete a single file. Missing file is success (mirrors rm -f). */
int fo_c_rm_file(const char *path) {
    if (!fo_has(path)) return 0;
    if (unlink(path) != 0 && errno != ENOENT) return -1;
    return 0;
}

/* mkdir -p: create path and all missing parents. */
int fo_c_mkdir_p(const char *path) {
    char clean[PATH_MAX];
    char parent[PATH_MAX];
    char *slash;
    struct stat st;
    size_t len, plen;

    if (!fo_has(path)) return -1;
    len = strlen(path);
    while (len > 1 && path[len - 1] == '/') len--;
    if (len >= sizeof(clean)) return -1;
    memcpy(clean, path, len);
    clean[len] = '\0';
    if (strcmp(clean, "/") == 0) return 0;
    if (stat(clean, &st) == 0) return S_ISDIR(st.st_mode) ? 0 : -1;

    slash = strrchr(clean, '/');
    if (slash != NULL && slash != clean) {
        plen = (size_t)(slash - clean);
        if (plen >= sizeof(parent)) return -1;
        memcpy(parent, clean, plen);
        parent[plen] = '\0';
        if (fo_c_mkdir_p(parent) != 0) return -1;
    }
    if (mkdir(clean, 0777) != 0 && errno != EEXIST) return -1;
    return 0;
}

/* Delete every regular file under root whose name ends with suffix. When
   recursive is nonzero, descends subdirectories (replacing find -name -delete).
   Returns the number of files removed, or -1 on a hard error. */
int fo_c_delete_suffix(const char *root, const char *suffix, int recursive) {
    DIR *dir;
    struct dirent *ent;
    char child[PATH_MAX];
    struct stat st;
    size_t slen, nlen;
    int removed = 0, sub;

    if (!fo_has(root) || !fo_has(suffix)) return 0;
    dir = opendir(root);
    if (dir == NULL) return (errno == ENOENT) ? 0 : -1;
    slen = strlen(suffix);
    while ((ent = readdir(dir)) != NULL) {
        if (strcmp(ent->d_name, ".") == 0 || strcmp(ent->d_name, "..") == 0)
            continue;
        if (snprintf(child, sizeof(child), "%s/%s", root, ent->d_name) >=
            (int)sizeof(child)) {
            continue;
        }
        if (lstat(child, &st) != 0) continue;
        if (S_ISDIR(st.st_mode)) {
            if (recursive) {
                sub = fo_c_delete_suffix(child, suffix, recursive);
                if (sub >= 0) removed += sub;
            }
            continue;
        }
        nlen = strlen(ent->d_name);
        if (nlen >= slen && strcmp(ent->d_name + (nlen - slen), suffix) == 0) {
            if (unlink(child) == 0) removed++;
        }
    }
    closedir(dir);
    return removed;
}

static int fo_str_contains(const char *hay, const char *needle) {
    if (needle == NULL || needle[0] == '\0') return 1;
    return strstr(hay, needle) != NULL;
}

/* Recursively collect regular files under root whose basename contains infix
   and ends with suffix, and whose full path contains path_needle (when set).
   Matches are written to out as NUL-separated paths; returns the count, or -1
   if the buffer overflows or a hard error occurs. Replaces a find pipeline. */
static int fo_collect_rec(const char *root, const char *infix,
                          const char *suffix, const char *path_needle,
                          int recursive, char *out, int cap, int *used) {
    DIR *dir;
    struct dirent *ent;
    char child[PATH_MAX];
    struct stat st;
    size_t nlen, slen, plen;
    int count = 0, sub;

    dir = opendir(root);
    if (dir == NULL) return (errno == ENOENT) ? 0 : 0;
    slen = (suffix != NULL) ? strlen(suffix) : 0;
    while ((ent = readdir(dir)) != NULL) {
        if (strcmp(ent->d_name, ".") == 0 || strcmp(ent->d_name, "..") == 0)
            continue;
        if (snprintf(child, sizeof(child), "%s/%s", root, ent->d_name) >=
            (int)sizeof(child)) {
            continue;
        }
        if (lstat(child, &st) != 0) continue;
        if (S_ISDIR(st.st_mode)) {
            if (!recursive) continue;
            sub = fo_collect_rec(child, infix, suffix, path_needle, recursive,
                                 out, cap, used);
            if (sub < 0) { closedir(dir); return -1; }
            count += sub;
            continue;
        }
        if (!S_ISREG(st.st_mode)) continue;
        nlen = strlen(ent->d_name);
        if (slen > 0 && (nlen < slen ||
            strcmp(ent->d_name + (nlen - slen), suffix) != 0)) continue;
        if (!fo_str_contains(ent->d_name, infix)) continue;
        if (!fo_str_contains(child, path_needle)) continue;
        plen = strlen(child);
        if (*used + (int)plen + 1 > cap) { closedir(dir); return -1; }
        memcpy(out + *used, child, plen);
        out[*used + (int)plen] = '\0';
        *used += (int)plen + 1;
        count++;
    }
    closedir(dir);
    return count;
}

int fo_c_collect_files(const char *root, const char *infix, const char *suffix,
                       const char *path_needle, int recursive, char *out,
                       int cap) {
    int used = 0;
    if (!fo_has(root)) return 0;
    return fo_collect_rec(root, infix, suffix, path_needle, recursive, out, cap,
                          &used);
}

/* Atomic exclusive directory create, used as a cross-process lock: returns 0
   when this caller created the directory, 1 when it already existed, -1 on a
   hard error. */
int fo_c_mkdir_excl(const char *path) {
    if (!fo_has(path)) return -1;
    if (mkdir(path, 0777) == 0) return 0;
    if (errno == EEXIST) return 1;
    return -1;
}

#include <signal.h>
/* Return 1 if a process with this pid exists, 0 otherwise (kill -0). */
int fo_c_pid_alive(int pid) {
    if (pid <= 0) return 0;
    if (kill((pid_t)pid, 0) == 0) return 1;
    return (errno == EPERM) ? 1 : 0;
}

/* File modification fingerprint for cache "outputs already match" checks:
   nanosecond mtime and byte size. Returns 0 on success, -1 if the path
   cannot be stat'd. Lets the build skip rewriting a large unchanged output
   (e.g. a 14MB statically linked binary) without re-hashing its contents. */
int fo_c_stat_fingerprint(const char *path, long long *mtime_ns,
                          long long *size) {
    struct stat st;
    if (!fo_has(path) || stat(path, &st) != 0) return -1;
#if defined(__APPLE__)
    *mtime_ns = (long long)st.st_mtimespec.tv_sec * 1000000000LL +
                (long long)st.st_mtimespec.tv_nsec;
#else
    *mtime_ns = (long long)st.st_mtim.tv_sec * 1000000000LL +
                (long long)st.st_mtim.tv_nsec;
#endif
    *size = (long long)st.st_size;
    return 0;
}

#include <time.h>
/* Sleep for the given milliseconds (no shell `sleep`). */
void fo_c_sleep_ms(int ms) {
    struct timespec ts;
    if (ms <= 0) return;
    ts.tv_sec = ms / 1000;
    ts.tv_nsec = (long)(ms % 1000) * 1000000L;
    nanosleep(&ts, NULL);
}

/* Recursively collect the unique parent directories of every *.mod file under
   root, NUL-separated in out. Returns the count, or -1 on overflow. Replaces
   `find -name '*.mod' -printf '%h\n' | sort -u`. */
static int fo_already_listed(const char *out, int used, const char *dirpath) {
    int i = 0;
    while (i < used) {
        if (strcmp(out + i, dirpath) == 0) return 1;
        i += (int)strlen(out + i) + 1;
    }
    return 0;
}

static int fo_collect_mod_dirs_rec(const char *root, char *out, int cap,
                                   int *used) {
    DIR *dir;
    struct dirent *ent;
    char child[PATH_MAX];
    struct stat st;
    size_t nlen, rlen;
    int count = 0, sub, have_mod = 0;

    dir = opendir(root);
    if (dir == NULL) return (errno == ENOENT) ? 0 : 0;
    while ((ent = readdir(dir)) != NULL) {
        if (strcmp(ent->d_name, ".") == 0 || strcmp(ent->d_name, "..") == 0)
            continue;
        if (snprintf(child, sizeof(child), "%s/%s", root, ent->d_name) >=
            (int)sizeof(child)) {
            continue;
        }
        if (lstat(child, &st) != 0) continue;
        if (S_ISDIR(st.st_mode)) {
            sub = fo_collect_mod_dirs_rec(child, out, cap, used);
            if (sub < 0) { closedir(dir); return -1; }
            count += sub;
            continue;
        }
        nlen = strlen(ent->d_name);
        if (nlen >= 4 && strcmp(ent->d_name + (nlen - 4), ".mod") == 0)
            have_mod = 1;
    }
    closedir(dir);
    if (have_mod && !fo_already_listed(out, *used, root)) {
        rlen = strlen(root);
        if (*used + (int)rlen + 1 > cap) return -1;
        memcpy(out + *used, root, rlen);
        out[*used + (int)rlen] = '\0';
        *used += (int)rlen + 1;
        count++;
    }
    return count;
}

int fo_c_collect_mod_dirs(const char *root, char *out, int cap) {
    int used = 0;
    if (!fo_has(root)) return 0;
    return fo_collect_mod_dirs_rec(root, out, cap, &used);
}

/* Copy src to dst (truncating dst), setting dst's mode to 0755 so an installed
   binary stays executable. Replaces cp -f for the install path. */
int fo_c_copy_exec(const char *src, const char *dst) {
    FILE *in, *out;
    char buf[65536];
    size_t n;

    if (!fo_has(src) || !fo_has(dst)) return -1;
    in = fopen(src, "rb");
    if (in == NULL) return -1;
    out = fopen(dst, "wb");
    if (out == NULL) { fclose(in); return -1; }
    while ((n = fread(buf, 1, sizeof(buf), in)) > 0) {
        if (fwrite(buf, 1, n, out) != n) { fclose(in); fclose(out); return -1; }
    }
    fclose(in);
    fclose(out);
    if (chmod(dst, 0755) != 0) return -1;
    return 0;
}

/* Rename src to dst (atomic within a filesystem). Replaces mv -f. */
int fo_c_rename_path(const char *src, const char *dst) {
    if (!fo_has(src) || !fo_has(dst)) return -1;
    if (rename(src, dst) != 0) return -1;
    return 0;
}

/* Append the bytes of src onto dst (mirrors cat src >> dst). */
int fo_c_append_file(const char *src, const char *dst) {
    FILE *in, *out;
    char buf[65536];
    size_t n;

    if (!fo_has(src) || !fo_has(dst)) return -1;
    in = fopen(src, "rb");
    if (in == NULL) return -1;
    out = fopen(dst, "ab");
    if (out == NULL) { fclose(in); return -1; }
    while ((n = fread(buf, 1, sizeof(buf), in)) > 0) {
        if (fwrite(buf, 1, n, out) != n) { fclose(in); fclose(out); return -1; }
    }
    fclose(in);
    fclose(out);
    return 0;
}
